tests_async/test_aiohttp_req.py

```diff
@@ -0,0 +1,226 @@
+
+        # await authed_session.close()
+
+    @pytest.mark.asyncio
+    async def test_request(self):
+        with aioresponses() as mocked:
+            credentials = mock.Mock(wraps=CredentialsStub())
+
+            mocked.get(self.TEST_URL, status=200, body="test")
+            session = aiohttp_req.AuthorizedSession(credentials)
+            resp = await session.request("GET", "http://example.com/")
+
+            assert resp.status == 200
+            assert "test" == await resp.text()
+
+            await session.close()
+
+    @pytest.mark.asyncio
+    async def test_ctx(self):
+        with aioresponses() as mocked:
+            credentials = mock.Mock(wraps=CredentialsStub())
+            mocked.get("http://test.example.com", payload=dict(foo="bar"))
+            session = aiohttp_req.AuthorizedSession(credentials)
+            resp = await session.request("GET", "http://test.example.com")
+            data = await resp.json()
+
+            assert dict(foo="bar") == data
+
+            await session.close()
+
+    @pytest.mark.asyncio
+    async def test_http_headers(self):
+        with aioresponses() as mocked:
+            credentials = mock.Mock(wraps=CredentialsStub())
+            mocked.post(
+                "http://example.com",
+                payload=dict(),
+                headers=dict(connection="keep-alive"),
+            )
+
+            session = aiohttp_req.AuthorizedSession(credentials)
+            resp = await session.request("POST", "http://example.com")
+
+            assert resp.headers["Connection"] == "keep-alive"
+
+            await session.close()
+
+    @pytest.mark.asyncio
+    async def test_regexp_example(self):
+        with aioresponses() as mocked:
+            credentials = mock.Mock(wraps=CredentialsStub())
+            mocked.get("http://example.com", status=500)
+            mocked.get("http://example.com", status=200)
+
+            session1 = aiohttp_req.AuthorizedSession(credentials)
+
+            resp1 = await session1.request("GET", "http://example.com")
+            session2 = aiohttp_req.AuthorizedSession(credentials)
+            resp2 = await session2.request("GET", "http://example.com")
+
+            assert resp1.status == 500
+            assert resp2.status == 200
+
+            await session1.close()
+            await session2.close()
+
+    @pytest.mark.asyncio
+    async def test_request_no_refresh(self):
+        credentials = mock.Mock(wraps=CredentialsStub())
+        with aioresponses() as mocked:
+            mocked.get("http://example.com", status=200)
+            authed_session = google.auth.transport.aiohttp_req.AuthorizedSession(
+                credentials
+            )
+            response = await authed_session.request("GET", "http://example.com")
+            assert response.status == 200
+            assert credentials.async_before_request.called
+            assert not credentials.refresh.called
+
+            await authed_session.close()
```

Comment:
Is closing the session required for cleanup? If so, what happens when an assertion fails? Is there a way to handle this via context managers or some pytest wizardry?

Reply:
Yes, the closing session was required to prevent the warning message saying that there is a loose session that is not closed. The tests still pass when we remove this, but was done to remove the warning message. If the assertion fails then an error is thrown before it reaches the end of the method (and hence the session.close line), so I assumed that would be okay?

Reply:
It doesn't seem hugely dangerous, however in python typically if there's a resource that needs to be closed after use we'd use a context manager to create it and then do necessary cleanup when the context is done. In unit tests this can often be achieved with some kind of pytest fixture. @crwilcox may have ideas about the best way to handle this.

Reply:
I think @tritone  is on the right track. We *can* do it the way authored, but usually as some kind of last resort. context managers are more pythonic and should be used when they can. Happy to chat about this if helpful.

Reply:
Sounds good, I can change the docstring to have this implementation instead. As I define the client session in the init function though, would we still use context managers in the implementation of the auth class?